docs/technical-design/file-upload.md

```diff
@@ -0,0 +1,93 @@
+- **The file is scanning.** After upload succeeds,  scanning begins automatically in AWS S3. The scanning indicator is displayed.
+- **The file processing is complete.** When scanning is successful (we use polling on the backend to determine if scanning is finished), we update the file item one final time in the list. That row will display as completed.
+- **The file reference is saved in the database**. When the user hits “continue” or “save as draft”  on the form, we store a reference to that file in the submission.
+
+### Possible error paths
+
+- The user selects or drops a file.
+  - If the user tries to drop a `.png` or some other unsupported image type, an **invalid file type error** displays.
+  - If the user selects or drops a file that is already present in the list, we display a **duplicate file error**.
+- The file is uploading. If uploading fails, we display an **upload failed error**.
+- The file is scanning. If scanning fails, we display a **scanning failed error**.
+- The user tries to hit “continue”.
+  - If scanning or uploading is not completed, we display a you must **wait to finish loading error**
+  - If there is a duplicate name error or some other error present on any file in the list, we display a **remove documents with issues** error.
+- It is possible for the user to have multiple types of errors in combination since we allow multiple files to be added at a time.
+
+### Limitations
+
+- FileUpload UI is optimized for a small list of files. For a larger number of files (15+) we likely want to move to a different UI. 
+  - This was explored in design iteration during the first year of the project. Options considered included a pop up toast menu that handles file upload outside the form (similar to Google Docs experience) and a file management page (similar to Box) with a large table or card view where users upload files, address any issues, and organize files outside the form.
+- Validations introduce complexity to the parent form component.  
+  - FileUpload related validations are hard to test and can overload form pages. The longest most complex tests in the codebase relate to forms where there are multiple file uploads (rates details- where multiple rate upload components may be loaded). This leads to performance issues and clunky UI since the user could be far scrolled down in the form but be blocked by an async file error that just appeared higher up on the form outside the viewport.
+- Difficult to load and performance test. We don’t currently have an easy way to see how the FileUpload would handle 20 large files on a low bandwidth connection, for example.
+- Slow overall upload/scan time (~30 seconds per file from user select to file complete). This impacts user experience as well as end to end and manual testing. Some of the slowness comes from our inherited tooling and could be addressed with tech effort.
```

Comment:
@haworku  again regarding accessibility, does this slowness also affect the screen reader experience in terms of perceivability  and real time feedback? That might be worth noting.